Sources/llbuild2/Core/Engine.swift

```diff
@@ -73,28 +77,59 @@ public protocol LLBFunction {
     func compute(key: LLBKey, _ fi: LLBFunctionInterface) -> LLBFuture<LLBValue>
 }
 
+public protocol LLBFunctionCache {
+    func get(key: LLBKey) -> LLBFuture<LLBDataID?>
+    func update(key: LLBKey, value: LLBDataID) -> LLBFuture<Void>
+}
+
 open class LLBTypedCachingFunction<K: LLBKey, V: LLBValue>: LLBFunction {
     public init() {}
 
+    private func computeAndUpdate(key: LLBKey, typedKey: K, _ fi: LLBFunctionInterface) -> LLBFuture<LLBValue> {
```

Comment:
is there a reason you need both versions of the same key? can you do `typedKey as LLBKey` when you need the LLBKey type?

Reply:
oh, maybe